### ISSUE TITLE
Refactor of Trilinos Recipe (Enable/Disable Optional)

### ISSUE THREAD

**Issue by kuberry** [kuberry/spack#1]
<h3>Current Approach</h3>
Setting `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` to `OFF`

<h4>Pros</h4>

    1.) you knew what was enabled by looking at the spec 
    2.) you approximately knew what was enabled by looking at how you called the recipe (defaults preset to `ON` are subject to change and were originally based on a particular application.
    3.) consistent with Spack allowing variant to be ON or OFF, with one of these values taken default.

<h4>Cons</h4>
  
    1.) because of concretizer output, conflicts and dependencies are not constructive or informative (requires reading the Spack recipe itself) 
    2.) most Trilinos package dependencies and conflicts between packages are not captured currently 
    3.) most Trilinos package TPLs are not captured currently.
  
<h4>Experience:</h4>
Anyone who wants to build Trilinos in Spack must experiment with their spec by trying various `+variant`s until they get a working combination (which they will not know _a priori_ unless they read the Tribits dependency files for that version of Trilinos).

<h3>Alternate Approach</h3>
Setting `Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES` to `ON`

<h4>Pros</h4>

    1.) easy to build working Trilinos recipe calls (`+variant` will work by enabling optional dependencies of `variant` package in Trilinos)
    2.) consistent with how Trilinos is usually installed by users today
    3.) Tribits logic will be used to determine other Trilinos packages that are dependencies, and enable them if necessary
    4.) not necessary to capture dependencies/conflicts between Trilinos packages (unless you want the spec to accurately reflect what is enabled/disabled)

<h4>Cons</h4>

    1.) inconsistent with Spack allowing variant to be `ON` or `OFF`, with one of these values taken by default. Would likely require a `+disable_variant` that translate to `Trilinos_ENABLE_{variant}:BOOL=OFF` in CMake and that create a conflict for `+variant`, ensuring `~variant`. `+variant` would still enable a package, but `~variant` would not ensure that the package is disabled (it could be implicitly enabled by Tribits because it is some other package dependency)
    2.) Just because `~variant` is in the spec, doesn't mean that the variant is not enabled by Tribits, so TPL dependencies of packages cannot be captured with this approach without capturing Tribits logic in Spack or enabling all required TPLs by default.

<h4>Experience</h4>
Can quickly construct a Trilinos install that is as minimal as just Teuchos and as large as whatever is needed, e.g. `Trilinos+muelu` would install Trilinos with all Trilinos packages enabled that MueLu relies on. After the install, other recipe maintainers  can still experiment with adding `+disable_variant` to create a minimal build.

TPLs could be missing when installing.

<h2>Addition to alternate approach:</h2>

<h3>Capturing Package Requirements (Offline)</h3>

```sh
cmake -DTrilinos_DEPS_XML_OUTPUT_FILE=out.xml -P ../cmake/tribits/ci_support/TribitsDumpDepsXmlScript.cmake`
```

Create a parser that reads in the output of the above CMake generated XML files and produces the necessary `+variant` for packages that would be turned on by Tribits implicitly. Use the same output to generate TPL dependencies based on `+variant`. Start with a certain version of Trilinos, then with each additional release capture the deltas in the logic.

tldr; Current approach has incomplete inter-dependencies and TPL dependencies. Proposed approach resolves inter-dependencies but has incomplete TPL dependencies. Capturing Tribits dependencies offline could fix the TPL dependencies.

**Comment by kuberry** [comment]
Script to generate spack variants and constraints from Tribits is https://raw.githubusercontent.com/kuberry/spack/tribits_spack_generation/spack_variants_from_tribits_projects.py

**Comment by kuberry** [comment]
https://github.com/spack/spack/issues/577 indicates that `~` should really only be used to indicate an incompatibility with a feature, rather than simply not wanting a feature enabled (because of effects on concretization and reuse of installations).
